nova/scheduler/external.py

```diff
@@ -0,0 +1,85 @@
+# Copyright (c) 2025 SAP SE
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+"""
+This module provides functionality to interact with an external scheduler API
+to reorder and filter hosts based on additional criteria.
+
+The external scheduler API is expected to take a list of weighed hosts and
+their weights, along with the request specification, and return a reordered
+and filtered list of host names.
+"""
+from oslo_log import log as logging
+import requests
+
+from nova.scheduler import utils
+
+LOG = logging.getLogger(__name__)
+
+
+def call_external_scheduler_api(weighed_hosts, weights, spec_obj, url):
+    """Reorder and filter hosts using an external scheduler service."""
+    if not url or not weighed_hosts:
+        return weighed_hosts
+    json_data = {
+        "spec": spec_obj.obj_to_primitive(),
+        "rebuild": utils.request_is_rebuild(spec_obj),
+        # Only provide basic information for the hosts for now.
+        # The external scheduler is expected to fetch statistics
+        # about the hosts separately, so we don't need to pass
+        # them here.
+        "hosts": [
+            {
+                "host": h.host,  # e.g. nova-compute-bb123
+                "hypervisor_hostname": h.hypervisor_hostname,
+                "status": h.status,
+            } for h in weighed_hosts
+        ],
+        # Also pass previous weights from the Nova weigher pipeline.
+        # The external scheduler api is expected to take these weights
+        # into account if provided.
+        "weights": weights,
+    }
+    LOG.debug("Calling external scheduler API with %s", json_data)
+    try:
+        response = requests.post(url, json=json_data, timeout=10)
+        response.raise_for_status()
+    except requests.RequestException as e:
+        LOG.error("Failed to call external scheduler API: %s", e)
+        return weighed_hosts
+
+    # The external scheduler api is expected to return a json with
+    # a sorted list of host names. Note that no weights are returned.
+    # The response structure should look like: {"hosts": ["host1", "host2"]}
+    response_json = response.json()
+    if "hosts" not in response_json:
```

Comment:
In `nova-api`, we have JSON schema to validate the JSON we get from users (and if I saw it correctly in the future also for returning JSON from the API). Might be a good idea to create such a schema here and validate that we got something we fully understand.

Reply:
Thanks, looking into that. It's surely more robust to use a json schema validator here.